java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -123,60 +115,61 @@ public String toString() {
     return buffer.toString();
   }
 
-  protected RelationState resolveState(Collection<BinaryRelation> knownRelations) {
-    if(hasSameOperand()) {
+  protected RelationState resolveState(List<BinaryRelation> knownRelations) {
+    if (hasSameOperand()) {
       return relationStateForSameOperand();
     }
-    Set<BinaryRelation> usedRelations = new HashSet<>();
-    Set<BinaryRelation> relations = new HashSet<>(knownRelations);
-    int iterations = 0;
-    while (!relations.isEmpty()) {
-      if (usedRelations.size() > 200 || iterations > MAX_ITERATIONS) {
-        // safety mechanism in case of an error in the algorithm
-        // should not happen under normal conditions
-        throw new TransitiveRelationExceededException("Used relations: " + relations.size() + ". Iterations " + iterations);
+    List<BinaryRelation> allRelations = deduce(knownRelations);
+    for (BinaryRelation relation : allRelations) {
```

Comment:
I think this could leverage stream api : deduce().stream().map().filter().firstMatch().orElse...